### PR TITLE
rgw: fix memory leak in RGWGetObjLayout

### PR DESCRIPTION
**Pull request by Liuchang0812**
target was malloced in RGWGetObjLayout::execute(), We should free it in descontrcuted function.

Signed-off-by: liuchang0812 <liuchang0812@gmail.com>

### CODE REVIEW

**Comment by Liuchang0812** [comment]
@joscollin thanks. I have updated my code.

File: src/rgw/rgw_op.h [hunk 1]
```diff
@@ -1773,7 +1773,8 @@ class RGWGetObjLayout : public RGWOp {
   rgw_raw_obj head_obj;
 
 public:
-  RGWGetObjLayout() {
+  RGWGetObjLayout() = default;
+  ~RGWGetObjLayout() {
```
Comment (cbodley) [inline src/rgw/rgw_op.h]: looks correct. how about using `std::unique_ptr<RGWRados::Object> target;` instead though? then we don't need a constructor or destructor at all

Comment (joscollin) [inline src/rgw/rgw_op.h]: >> how about using std::unique_ptr<RGWRados::Object> target; instead though? then we don't need a constructor or destructor at all

It is good if target is used only inside RGWGetObjLayout::execute() and should be deleted when it leaves the scope. If the scope of target is with in the stack, then why allocating the memory using new operator in RGWGetObjLayout::execute() ?

Comment (joscollin) [inline src/rgw/rgw_op.h]: As I checked quickly, the member variable target is only used inside the execute() function. So no need of a member variable. Just use `std::unique_ptr<RGWRados::Object> target;` locally inside execute() with new operator and without a delete. That should be fine, unless the target variable is used somewhere else later.

Comment (cbodley) [inline src/rgw/rgw_op.h]: > Otherwise, avoid using new and create the object with in the stack is also fine.

even better 👍

Comment (Liuchang0812) [inline src/rgw/rgw_op.h]: Copy that

**Comment by Liuchang0812** [comment]
@cbodley @joscollin 

updated! thanks for your suggestions.

**Comment by Liuchang0812** [comment]
jenkins test this please

File: src/rgw/rgw_op.cc [hunk 2]
```diff
@@ -5526,8 +5526,8 @@ void RGWGetObjLayout::pre_exec()
 void RGWGetObjLayout::execute()
 {
   rgw_obj obj(s->bucket, s->object);
-  target = new RGWRados::Object(store, s->bucket_info, *static_cast<RGWObjectCtx *>(s->obj_ctx), rgw_obj(s->bucket, s->object));
-  RGWRados::Object::Read stat_op(target);
+  std::unique_ptr<RGWRados::Object> target(new RGWRados::Object(store, s->bucket_info, *static_cast<RGWObjectCtx *>(s->obj_ctx), rgw_obj(s->bucket, s->object)));
+  RGWRados::Object::Read stat_op(target.get());
```
Comment (cbodley) [inline src/rgw/rgw_op.cc]: is there a reason this can't be on the stack instead?

File: src/rgw/rgw_op.h [hunk 3]
```diff
@@ -1773,9 +1773,8 @@ class RGWGetObjLayout : public RGWOp {
   rgw_raw_obj head_obj;
 
 public:
-  RGWGetObjLayout() {
-    delete target;
-  }
+  RGWGetObjLayout() = default;
+  ~RGWGetObjLayout() = default;
```
Comment (cbodley) [inline src/rgw/rgw_op.h]: can remove these declarations altogether, the compiler will still generate them

also, please remove the unused member variable `RGWRados::Object *target{nullptr};`

Comment (joscollin) [inline src/rgw/rgw_op.h]: That's right. I have already gave those comments earlier, but I'm not able to see them in this PR now.

Comment (Liuchang0812) [inline src/rgw/rgw_op.h]: thanks @cbodley @joscollin 

I prefer use a stack variable.  I updated code.

**Comment by cbodley** [comment]
see build failure:
```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_op.cc:5542:18: error: base operand of ‘->’ has non-pointer type ‘RGWRados::Object’
   op_ret = target->get_manifest(&manifest);
```

**Comment by joscollin** [comment]
target is an object now, not a pointer.

**Comment by Liuchang0812** [comment]
sorry for that compile error, I will update and build/test it tomorrow, I don't have a dev-env PC now.

**Comment by Liuchang0812** [comment]
Ping

File: src/rgw/rgw_op.h [hunk 4]
```diff
@@ -1768,14 +1768,10 @@ class RGWSetAttrs : public RGWOp {
 
 class RGWGetObjLayout : public RGWOp {
 protected:
-  RGWRados::Object *target{nullptr};
   RGWObjManifest *manifest{nullptr};
   rgw_raw_obj head_obj;
 
 public:
-  RGWGetObjLayout() {
```
Comment (joscollin) [inline src/rgw/rgw_op.h]: Please don't remove something that is not related to this particular fix. Please don't remove the constructor.

Comment (joscollin) [inline src/rgw/rgw_op.h]: Yes, Your fix doesn't need to remove the already existing constructor. It can be used to initialize variables later.

Comment (Liuchang0812) [inline src/rgw/rgw_op.h]: updated!

**Comment by cbodley** [comment]
how is this being tested?

**Comment by Liuchang0812** [comment]
@cbodley  I will try to add a unittest for it and test the `RGWGetObjLayout` command by hands. I'm not familiar with `RGWGetObjLayout` op, may need your help。

**Comment by cbodley** [comment]
@Liuchang0812 okay. the op was added for the https://github.com/ceph/ragweed testing framework. i think that testing it by hand with a curl command or similar is okay for this pr

**Comment by Liuchang0812** [comment]
@cbodley  It's ok with manual test.
```
➜  build git:(wip-fix-seg-fault) ✗ curl -v 'http://localhost:8000/liuchang/make.txt?layout'
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 8000 (#0)
> GET /liuchang/make.txt?layout HTTP/1.1
> Host: localhost:8000
> User-Agent: curl/7.47.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< x-amz-request-id: tx000000000000000000002-0058d38fb1-ac49-default
< Content-Type: application/json
< Content-Length: 869
< Date: Thu, 23 Mar 2017 09:04:49 GMT
< 
* Connection #0 to host localhost left intact
{"head":{"pool":"default.rgw.buckets.data","oid":"1afed874-2d67-4188-bec1-776d7b796389.4158.1_make.txt","loc":""},"manifest":{"objs":[],"obj_size":235844,"explicit_objs":"false","head_size":235844,"max_head_size":4194304,"prefix":".3rHVtojliPBXWY9R9lHJWRXHWCK8hvu_","rules":[{"key":0,"val":{"start_part_num":0,"start_ofs":4194304,"part_size":0,"stripe_max_size":4194304,"override_prefix":""}}],"tail_instance":"","tail_placement":{"bucket":{"name":"liuchang","marker":"1afed874-2d67-4188-bec1-776d7b796389.4158.1","bucket_id":"1afed874-2d67-4188-bec1-776d7b796389.4158.1","tenant":"","explicit_placement":{"data_pool":"","data_extra_pool":"","index_pool":""}},"placement_rule":"default-placement"}},"data_location":[{"ofs":0,"loc":{"pool":"default.rgw.buckets.data","oid":"1afed874-2d67-4188-bec1-776d7b796389.4158.1_make.txt","loc":""},"loc_ofs":0,"loc_size":235844}]}%   
```
